astropy/io/fits/fitsrec.py

```diff
@@ -1017,12 +1017,32 @@ def _get_heap_data(self):
 
         This is returned as a numpy byte array.
         """
-        if self._heapsize:
-            raw_data = self._get_raw_data().view(np.ubyte)
+        raw_data = self._get_raw_data()
+        if self._heapsize and raw_data is not None:
+            # Read the heap from disk
+            raw_data = raw_data.view(np.ubyte)
             heap_end = self._heapoffset + self._heapsize
             return raw_data[self._heapoffset : heap_end]
         else:
-            return np.array([], dtype=np.ubyte)
+            # Data is only in memory so create the heap data, one column
+            # at a time, in the order that the data pointers appear in the
+            # column (regardless if that data pointer has a different,
+            # previous heap offset listed)
+            data = []
+            for idx in range(self._nfields):
+                # data should already be byteswapped from the caller
+                # using _binary_table_byte_swap
+                if not isinstance(self.columns._recformats[idx], _FormatP):
+                    continue
+
+                for row in self.field(idx):
+                    if len(row) > 0:
+                        data.append(row.view(type=np.ndarray, dtype=np.ubyte))
+
+            if data:
```

Comment:
I'd go for
```
return np.concatenate(data) if data else np.array([], dtype=np.ubyte)
```
but, really, it is a total nitpick.

p.s.  I had really hoped this could be turned into a proper one-liner, but `concatenate` doesn't like empty lists, and `np.array` wants equal sized arrays. Pity.

Reply:
Both versions would be fine to me, so to avoid wasting human and CI time I will merge as is ;).
Thanks for the review ! :)